Adding a "HOW-TO: Building your own collector" tutorial to the collector (#1277)
open-telemetry/opentelemetry.io: content/en/docs/collector/howto-custom-collector.md

> # HOW-TO: Build your own Collector instance

Reviewer: Not a hugo expert and if this is mandatory, but most pages have a title like this:
```markdown
---
title: "Building your own collector instance"
weight: 30
---
```
Compare it with https://raw.githubusercontent.com/open-telemetry/opentelemetry.io/main/content/en/docs/collector/custom-auth.md -- I think there should also be some consistency in naming ("Building ..."

Author: Hi @svrnm, fixed on (7d3d01)[ https://github.com/open-telemetry/opentelemetry.io/pull/1277/commits/7d3d01bd356ff8c6ae72ffd52cc476f69108f7ca

Reviewer: @rquedas , thanks, you can "resolve" this comment then

> If you are planning to build and debug your components, you are going to need your own Collector instance. That will allow you to launch and debug your Collector components directly within your favorite Golang IDE.

Reviewer: say "OpenTelemetry Collector" at the first instance of "Collector"

Author: @svrnm, fixed on [7d3d01](https://github.com/open-telemetry/opentelemetry.io/pull/1277/commits/7d3d01bd356ff8c6ae72ffd52cc476f69108f7ca)

> The other really cool aspect of approaching the component development this way is that you can use all the cool debugging features from your IDE (stack traces are great teachers!) to understand how the Collector itself interacts with your component code.

Reviewer: Remove the second "cool"

Author: Hi @svrnm, fixed on [7d3d01](https://github.com/open-telemetry/opentelemetry.io/pull/1277/commits/7d3d01bd356ff8c6ae72ffd52cc476f69108f7ca)

> In order to me it easy for developers to create their own custom builds of the Collector, the OpenTelemetry Community has developed a tool called `OpenTelemetry Collector Builder`.

Reviewer 2: ```suggestion
In order to make it easy for developers to create their own custom builds of the Collector, the OpenTelemetry Community has developed a tool called `OpenTelemetry Collector Builder`.
```

> In order to make it easy for developers to create their own custom builds of the Collector, the OpenTelemetry Community has developed a tool called `OpenTelemetry Collector Builder`.

Reviewer: add a link to the repo/folder where the builder lives

Author: Hi @svrnm, fixed on [7d3d01](https://github.com/open-telemetry/opentelemetry.io/pull/1277/commits/7d3d01bd356ff8c6ae72ffd52cc476f69108f7ca)

> The `builder` is meant to help developers quickly assemble and build their own collector's distributions, based on a given configuration file.

Reviewer: check that you consistently write "collector" or "Collector", from a few other pages it looks like that it is "Collector"

Author: Hi @svrnm, fixed on [7d3d01](https://github.com/open-telemetry/opentelemetry.io/pull/1277/commits/7d3d01bd356ff8c6ae72ffd52cc476f69108f7ca)

> The `builder` is meant to help developers to quickly assemble and build their own collector's distributions, based on a given configuration file.

Reviewer 2: ```suggestion
The `builder` is meant to help developers quickly assemble and build their own collector's distributions, based on a given configuration file.
```

> As part of the process the `builder` will generate the collector's source code, which you can borrow and steal as your own to help build and debug your own components, so let's get started.

Reviewer 2: ```suggestion
As part of the process the `builder` will generate the collector's source code, which you can modify to help build and debug your own components, so let's get started.
```

## Step 1 - Install the builder

> You can easily install the builder through the go install command. 

Reviewer 2: ```suggestion
You can install the builder through the `go install` command. 
```

    name: otelcol-mybuild # the binary name. Optional.
    description: "Custom OpenTelemetry Collector distribution" # a long name for the application. Optional.
    otelcol_version: "0.41.0" # the OpenTelemetry Collector version to use as base for the distribution. Optional.
    output_path: ./otelcol-mybuild # the path to write the output (sources and binary). Optional.
    version: "1.0.0" # the version for your custom OpenTelemetry Collector. Optional.
    go: "/usr/local/go/bin/go" # which Go binary to use to compile the generated sources. Optional.
exporters:
  - gomod: "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/jaegerexporter v0.41.0"
  - import: go.opentelemetry.io/collector/exporter/loggingexporter
    gomod: go.opentelemetry.io/collector v0.41.0

receivers:
  - import: go.opentelemetry.io/collector/receiver/otlpreceiver
    gomod: go.opentelemetry.io/collector v0.41.0

processors:
  - import: go.opentelemetry.io/collector/processor/batchprocessor
    gomod: go.opentelemetry.io/collector v0.41.0
```

> For more information on how to create your builder config file, checkout [OpenTelemetry Collector builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder) folder inside the Collector's project in GitHub.

Reviewer 2: ```suggestion
For more information on how to create your builder config file, check out the [OpenTelemetry Collector builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder) folder inside the Collector's project in GitHub.
```

---
title: "Building your own Collector instance"
weight: 30
---

If you are planning to build and debug your components, you are going to need your own Collector instance. That will allow you to launch and debug your OpenTelemetry Collector components directly within your favorite Golang IDE.

The other interesting aspect of approaching the component development this way is that you can use all the debugging features from your IDE (stack traces are great teachers!) to understand how the Collector itself interacts with your component code.

In order to make it easy for developers to create their own custom builds of the Collector, the OpenTelemetry Community has developed a tool called [OpenTelemetry Collector builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder).

> The `builder` is meant to help developers to quickly assemble and build their own Collector's distributions, based on a given configuration file.

Reviewer 3: That's not the main goal for the builder. The builder exists to assist people in assembling their own distribution, making it easy to build a distribution that includes their custom components along with components that are publicly available.

The developer use-case is a nice added benefit though :-)

Author: @jpkrohling fair enough, let me fix that.

Author: fixed in: https://github.com/open-telemetry/opentelemetry.io/pull/1277#:~:text=to%20support%20suggestions.-,fd2e9a3,-Merge%20branch%20%27doc

As part of the process the `builder` will generate the Collector's source code, which you can borrow and steal as your own to help build and debug your own components, so let's get started.

## Step 1 - Install the builder

You can install the builder through the `go install` command. 

Open your terminal and type:

```

> go install go.opentelemetry.io/collector/cmd/builder@latest

Reviewer 3: This is not the recommended way to install the builder. From the builder's readme: https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder#installation

Author: fixed in: https://github.com/open-telemetry/opentelemetry.io/pull/1277#:~:text=to%20support%20suggestions.-,fd2e9a3,-Merge%20branch%20%27doc

```

> The `builder` command will be installed on your `$GOPATH/bin`, which most likely is already added to your path.

Reviewer 3: `ocb` is what should be available when following the official installation instructions.

Author: fixed in : https://github.com/open-telemetry/opentelemetry.io/pull/1277#:~:text=to%20support%20suggestions.-,fd2e9a3,-Merge%20branch%20%27doc

      --description string       A descriptive name for the OpenTelemetry Collector distribution (default "Custom OpenTelemetry Collector distribution")
      --go string                The Go binary to use during the compilation phase. Default: go from the PATH
  -h, --help                     help for builder
      --module string            The Go module for the new distribution (default "go.opentelemetry.io/collector/cmd/builder")
      --name string              The executable name for the OpenTelemetry Collector distribution (default "otelcol-custom")
      --otelcol-version string   Which version of OpenTelemetry Collector to use as base (default "0.42.0")
      --output-path string       Where to write the resulting files (default "/var/folders/86/s7l1czb16g124tng0d7wyrtw0000gn/T/otelcol-distribution3618633831")
      --skip-compilation         Whether builder should only generate go code with no compile of the collector (default false)
      --version string           The version for the OpenTelemetry Collector distribution (default "1.0.0")

Use "builder [command] --help" for more information about a command.
```

## Step 2 - Create a builder config file

> The builder config file is a `yaml` where you basically pass information about the code generation and compile process combined with the components that you would like to add to your Collector's distribution.

Reviewer 3: ```suggestion
The builder's manifest is a `yaml` where you basically pass information about the code generation and compile process combined with the components that you would like to add to your Collector's distribution.
```

> Go ahead and create a `builder-config.yaml` with the following content:

Reviewer 3: ```suggestion
Go ahead and create a `manifest.yaml` with the following content:
```

> builder-config.yaml
```yaml
dist:

>     module: github.com/open-telemetry/opentelemetry-collector # the module name for the new distribution, following Go mod conventions. Optional, but recommended.

Reviewer 3: This is the distribution's module name. It should not be using the same name as our official distributions.

Author: I missed that one, I will fix it in the next commit

    name: otelcol-mybuild # the binary name. Optional.
    description: "Custom OpenTelemetry Collector distribution" # a long name for the application. Optional.

>     otelcol_version: "0.41.0" # the OpenTelemetry Collector version to use as base for the distribution. Optional.

Reviewer 3: Use the latest as of now, please.

Author: got it, I will fix it in the next commit

Author: fixed in: https://github.com/open-telemetry/opentelemetry.io/pull/1277#:~:text=to%20support%20suggestions.-,fd2e9a3,-Merge%20branch%20%27doc

>     output_path: ./otelcol-mybuild # the path to write the output (sources and binary). Optional.

Reviewer 3: There's no need to override this without a good reason.

Author: @jpkrohling: which setting? all of them?

Reviewer 3: Just the `output_path`. The builder has reasonable defaults, so, most of the properties can be left out, but the set of properties you have here (except for this output_path) are good to have in the manifest.

Author: fixed in: https://github.com/open-telemetry/opentelemetry.io/pull/1277#:~:text=to%20support%20suggestions.-,fd2e9a3,-Merge%20branch%20%27doc

    version: "1.0.0" # the version for your custom OpenTelemetry Collector. Optional.

>     go: "/usr/local/go/bin/go" # which Go binary to use to compile the generated sources. Optional.

Reviewer 3: Same here.

Author: which setting?

Reviewer 3: `go`

Author: fixed in: https://github.com/open-telemetry/opentelemetry.io/pull/1277#:~:text=to%20support%20suggestions.-,fd2e9a3,-Merge%20branch%20%27doc

---
title: "Building your own Collector instance"
weight: 30
---

> If you are planning to build and debug your components, you are going to need your own Collector instance. That will allow you to launch and debug your OpenTelemetry Collector components directly within your favorite Golang IDE.

Reviewer 2: ```suggestion
If you are planning to build and debug custom collector receivers, processors, extensions, or exporters, you are going to need your own Collector instance. That will allow you to launch and debug your OpenTelemetry Collector components directly within your favorite Golang IDE.
```

The other interesting aspect of approaching the component development this way is that you can use all the debugging features from your IDE (stack traces are great teachers!) to understand how the Collector itself interacts with your component code.

The OpenTelemetry Community developed a tool called [OpenTelemetry Collector builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder)(aka `ocb`) to assist people in assembling their own distribution, making it easy to build a distribution that includes their custom components along with components that are publicly available.

> As part of the process the `builder` will generate the Collector's source code, which you can borrow and steal as your own to help build and debug your own components, so let's get started.

Reviewer 2: ```suggestion
As part of the process the `builder` will generate the Collector's source code, which you can use to help build and debug your own custom components, so let's get started.
```

      --description string       A descriptive name for the OpenTelemetry Collector distribution (default "Custom OpenTelemetry Collector distribution")
      --go string                The Go binary to use during the compilation phase. Default: go from the PATH
  -h, --help                     help for builder
      --module string            The Go module for the new distribution (default "go.opentelemetry.io/collector/cmd/builder")
      --name string              The executable name for the OpenTelemetry Collector distribution (default "otelcol-custom")
      --otelcol-version string   Which version of OpenTelemetry Collector to use as base (default "0.42.0")
      --output-path string       Where to write the resulting files (default "/var/folders/86/s7l1czb16g124tng0d7wyrtw0000gn/T/otelcol-distribution3618633831")
      --skip-compilation         Whether builder should only generate go code with no compile of the collector (default false)
      --version string           The version for the OpenTelemetry Collector distribution (default "1.0.0")

Use "builder [command] --help" for more information about a command.
```

## Step 2 - Create a builder manifest file

> The builder's `manifest` file is a `yaml` where you basically pass information about the code generation and compile process combined with the components that you would like to add to your Collector's distribution.

Reviewer 2: ```suggestion
The builder's `manifest` file is a `yaml` where you pass information about the code generation and compile process combined with the components that you would like to add to your Collector's distribution.
```

> The `manifest` starts with a map named `dist` which basically contains tags to help you configure the code generation and compile process. In fact, all the tags for `dist` are the equivalent of the `ocb` command line `flags`.

Reviewer 2: ```suggestion
The `manifest` starts with a map named `dist` which contains tags to help you configure the code generation and compile process. In fact, all the tags for `dist` are the equivalent of the `ocb` command line `flags`.
```

Here are the tags for the `dist` map:

| Tag              | Description | Optional | Default Value|
|------------------|-------------|----------|--------------|
| module:          | The module name for the new distribution, following Go mod conventions. Optional, but recommended.| Yes | "go.opentelemetry.io/collector/cmd/builder" |
| name:            | The binary name for your distribution | Yes | "otelcol-custom" |
| description:     | A long name for the application. | Yes | "Custom OpenTelemetry Collector distribution" |

> | otelcol_version: | The OpenTelemetry Collector version to use as base for the distribution. | Yes | "0.49.0" | 

Reviewer 2: ```suggestion
| otelcol_version: | The OpenTelemetry Collector version to use as base for the distribution. | Yes | "0.53.0" | 
```

As you can see on the table above, all the `dist` tags are optional, so you will be adding custom values for them depending if your intentions to make your custom Collector distribution available for consumption by other users or if you are simply leveraging the `ocb` to bootstrap your component development and testing environment.

For this tutorial, you will be creating a Collector's distribution to support the development and testing of components.

Go ahead and create a manifest file named `builder-config.yaml` with the following content:

> builder-config.yaml
```yaml
dist:
    name: otelcol-dev 
    description: "Basic OTel Collector distribution for Developers"
    output_path: ./otelcol-dev 
```

Now you need to add the modules representing the components you want to be incorporated in this custom Collector distribution. Take a look at the [ocb configuration documentation](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder#configuration) to understand the different modules and how to add the components.

We will be adding the following components to our development and testing collector distribution:

- Exporters: Jaeger and Logging

> - Receivers: Otlp

Reviewer 2: ```suggestion
- Receivers: OTLP
```
